.github/workflows/hash-backend-cd.yml

```diff
@@ -0,0 +1,139 @@
+  HASH_SEARCHLOADER_RESOURCE_NAME: ${{ secrets.HASH_SEARCHLOADER_RESOURCE_NAME }}
+
+  HASH_ECS_CLUSTER_NAME: ${{ secrets.HASH_ECS_CLUSTER_NAME }}
+
+name: HASH backend deployment
+jobs:
+  build-api:
+    name: Build and push HASH api image
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+
+      - uses: ./.github/actions/docker-ecr-login
+        with:
+          AWS_ACCESS_KEY_ID: ${{ env.AWS_ACCESS_KEY_ID }}
+          AWS_SECRET_ACCESS_KEY: ${{ env.AWS_SECRET_ACCESS_KEY }}
+          AWS_REGION: ${{ env.AWS_REGION }}
+
+      # Build images and push to ECR
+      - name: Build and push api
+        uses: docker/build-push-action@v2
+        env:
+          HASH_API_IMAGE_NAME: ${{ env.HASH_API_RESOURCE_NAME }}ecr
+        with:
+          # The Dockerfile is in a leaf-folder without access to what it needs.
+          # Thus the context is within the root, while the dockerfile is deeply nested.
+          context: ${{ github.workspace }}
+          push: true
+          file: ${{ github.workspace }}/packages/hash/docker/api/prod/Dockerfile
+          tags: ${{ env.AWS_ECR_URL }}/${{ env.HASH_API_IMAGE_NAME }}:latest
+
+  build-realtime:
+    name: Build and push HASH realtime image
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+
+      - uses: ./.github/actions/docker-ecr-login
+        with:
+          AWS_ACCESS_KEY_ID: ${{ env.AWS_ACCESS_KEY_ID }}
+          AWS_SECRET_ACCESS_KEY: ${{ env.AWS_SECRET_ACCESS_KEY }}
+          AWS_REGION: ${{ env.AWS_REGION }}
+
+      - name: Build and push realtime
+        uses: docker/build-push-action@v2
+        env:
+          HASH_REALTIME_IMAGE_NAME: ${{ env.HASH_REALTIME_RESOURCE_NAME }}ecr
+        with:
+          context: ${{ github.workspace }}
+          push: true
+          file: ${{ github.workspace }}/packages/hash/docker/realtime/prod/Dockerfile
+          tags: ${{ env.AWS_ECR_URL }}/${{ env.HASH_REALTIME_IMAGE_NAME }}:latest
+          cache-from: type=local,src=/tmp/.buildx-cache
+          cache-to: type=local,dest=/tmp/.buildx-cache-new
+
+  build-searchloader:
+    name: Build and push HASH searchloader image
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+
+      - uses: ./.github/actions/docker-ecr-login
+        with:
+          AWS_ACCESS_KEY_ID: ${{ env.AWS_ACCESS_KEY_ID }}
+          AWS_SECRET_ACCESS_KEY: ${{ env.AWS_SECRET_ACCESS_KEY }}
+          AWS_REGION: ${{ env.AWS_REGION }}
+
+      - name: Build and push search-loader
+        uses: docker/build-push-action@v2
+        env:
+          HASH_SEARCHLOADER_IMAGE_NAME: ${{ env.HASH_SEARCHLOADER_RESOURCE_NAME }}ecr
+        with:
+          context: ${{ github.workspace }}
+          push: true
+          file: ${{ github.workspace }}/packages/hash/docker/search-loader/prod/Dockerfile
+          tags: ${{ env.AWS_ECR_URL }}/${{ env.HASH_SEARCHLOADER_IMAGE_NAME }}:latest
```

Comment:
What do you think about using a build ID as an additional tag which would let us identify which github actions workflow triggered the build?

We could either create a UUID in the workflow, or use the `GITHUB_RUN_ID` or `GITHUB_SHA` env vars - these wouldn't let us identify the exact run though. Not sure if I've missed a more appropriate env var from [this list](https://docs.github.com/en/actions/learn-github-actions/environment-variables)

Reply:
That might be a good idea. 
One thing is that it would require us to change the lifecycle policy we've added to the ECR repositories https://github.com/hashintel/internal/pull/3300, such that we are using a tag prefix instead of using `untagged`. I don't know if we can adapt our rule to conform to that without potentially evicting "latest" tag

Reply:
Good point, perhaps we can switch to a lifecycle policy that only keeps a certain number of images around - perhaps something like 5 - 10. That would allow us to rollback a couple of images, save storage costs, and wouldn't risk evicting the `latest` image.

Reply:
Yeah, that is possible, but then we could technically still end up with `latest` being evicted if we push images manually, right? I just don't want us to end up in a situation ever where latest is removed because of the lifecycle policy

Reply:
We could help ensure that by having a life-cycle rule that ensures there's always one or more images with the `latest` tag, and give it that lifecycle rule highest priority. If I'm understanding correctly that would prevent the `latest` image from ever being evicted.

Based on [this SO answer](https://stackoverflow.com/questions/51375318/ecr-lifecycle-policy-exception), I think the life-cycle rule would look something like:
```TypeScript
{
    "rules": [
        {
            "rulePriority": 1,
            "description": "Keep Latest Image",
            "selection": {
                "tagStatus": "tagged",
                "tagPrefixList": ["latest"],
                "countType": "imageCountMoreThan",
                "countNumber": 1
            },
            "action": {
                "type": "expire"
            }
        },
        {
            "rulePriority": 2,
            "description": "Prune Old Images",
            "selection": {
                "tagStatus": "any",
                "countType": "imageCountMoreThan",
                "countNumber": 10
            },
            "action": {
                "type": "expire"
            }
        }
    ]
}
```

Reply:
This is great, thank you. I just tested the policy and it works :)

I'll adapt this and change the pipeline

Reply:
Change is here https://github.com/hashintel/internal/pull/3301

Reply:
Also, I am aware that we have some duplication in the pipeline currently, which is not nice to keep track of. The work to add caching addresses this.

Reply:
![image](https://user-images.githubusercontent.com/6988668/166697197-55aaca89-b6d0-48c5-8d65-3912c9ae0311.png)
:+1: